Initial mqtt port

jvalkeal opened the pull request:
- Porting mqtt module as source and sink apps.
- Fixes spring-cloud-stream-app-starters/app-starters-release#8
- Fixes spring-cloud-stream-app-starters/app-starters-release#13

jvalkeal committed:
Initial mqtt port

- Porting mqtt module as source and sink apps.
- Fixes spring-cloud-stream-app-starters/app-starters-release#8
- Fixes spring-cloud-stream-app-starters/app-starters-release#13

jvalkeal committed:
Remove README.md

garyrussell commented on README.adoc:
```diff
@@ -0,0 +1,9 @@
+include::spring-cloud-starter-stream-source-mqtt/README.adoc[]
+
+## Rabbit Sink
+include::spring-cloud-starter-stream-sink-mqtt/README.adoc[]
```
`s/Rabbit/MQTT/` (both titles)

garyrussell commented on mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java:
```diff
@@ -0,0 +1,132 @@
+	/**
+	 * location of the mqtt broker(s) (comma-delimited list)
+	 */
+	private String url = "tcp://localhost:1883";
```
This should be `String[]` - Spring will parse the CDL - the factory setter takes `String[]` (varargs).

garyrussell commented on mqtt-app-starters-test-support/src/main/java/org/springframework/cloud/stream/app/test/mqtt/MqttRule.java:
```diff
@@ -0,0 +1,62 @@
+			connectionFactory.setConnectionTimeout(1);
+			client = connectionFactory.getClientInstance("tcp://localhost:1883", "scdf-test-client");
+			client.connect(connectionFactory.getConnectionOptions());
+		} catch (Exception e) {
```
catch on new line.

garyrussell commented on mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttConfiguration.java:
```diff
@@ -0,0 +1,53 @@
+		factory.setKeepAliveInterval(mqttProperties.getKeepAliveInterval());
+		if (ObjectUtils.nullSafeEquals(mqttProperties.getPersistence(), "file")) {
+			factory.setPersistence(new MqttDefaultFilePersistence(mqttProperties.getPersistenceDirectory()));
+		} else if (ObjectUtils.nullSafeEquals(mqttProperties.getPersistence(), "memory")) {
```
else on new line

garyrussell commented on mqtt-app-starters-test-support/src/main/java/org/springframework/cloud/stream/app/test/mqtt/MqttRule.java:
```diff
@@ -0,0 +1,62 @@
+ * @author Janne Valkealahti
+ *
+ */
+public class MqttRule extends TestWatcher implements TestRule {
```
For consistency with other starters, this should really extend `AbstractExternalResourceTestSupport`.

sobychacko commented:
We might want to set the version of this project at 1.0.0.BUILD-SNAPSHOT as this is a new addition to app starters.

jvalkeal commented:
@sobychacko This versioning is one of a weirdest things in apps. While it's true that `1.0.0` is probably more appropriate but as all apps currently have same version we haven't followed this approach that new app starts from `1.0.0`. Or is it a case that since initial `1.0.0` release we haven't added a singe new app?

jvalkeal committed:
Address review comments

- Use real MqttTestSupport
- Polish

sobychacko commented:
@jvalkeal we had new apps in bacon, but because of some special circumstances, we needed to baseline everything to 1.2.0. But i think going forward, for new app starters, we should start from 1.0.0.

artembilan commented on README.adoc:
```diff
@@ -0,0 +1,9 @@
+# MQTT Source and Sink
+
+To learn more about this application and the supported properties, please review the following link.
```
> Source and Sink

So, might be better to tell like: " these applications "

artembilan commented on mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttConfiguration.java:
```diff
@@ -0,0 +1,54 @@
+ * @author Janne Valkealahti
+ *
+ */
+public class MqttConfiguration {
```
Why this isn't marked with the `@Configuration`?
Does it really work as is?

jvalkeal replied:
If it's imported `@Configuration' is not needed.

jvalkeal replied:
I can add it, makes it less confusing.

artembilan commented on mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java:
```diff
@@ -0,0 +1,132 @@
+	/**
+	 * location of the mqtt broker(s) (comma-delimited list)
+	 */
+	private String[] url = new String[] { "tcp://localhost:1883" };
```
Some of the properties might be required - need particular validations to protect against misconfiguration

jvalkeal replied:
We used validators in xd to fail fast but it looks like we dropped using those. Not sure what to do?

artembilan replied:
Not sure what you mean about "dropped", but even mentioned `RabbitSourceProperties` has validation annotations

artembilan commented on spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkConfiguration.java:
```diff
@@ -0,0 +1,65 @@
+	@ServiceActivator(inputChannel = Sink.INPUT)
+	public MessageHandler mqttOutbound() {
+		MqttPahoMessageHandler messageHandler = new MqttPahoMessageHandler(properties.getClientId(), mqttClientFactory);
+		messageHandler.setAsync(true);
```
`this.properties.isAsync()` must be used.

jvalkeal replied:
Good catch, I'll need to add that.

artembilan commented on spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java:
```diff
@@ -0,0 +1,106 @@
+ * @author Janne Valkealahti
+ *
+ */
+@ConfigurationProperties("mqtt")
```
I wonder how does that work if we use the same prefix on different `@ConfigurationProperties` in the same app...
I mean that `@EnableConfigurationProperties({ MqttProperties.class, MqttSinkProperties.class })` and this `mqtt` prefix on the `MqttProperties` as well. 

Does it mean that Boot properly merge them into one set and then is able to distribute provided properties to proper classes?
Just confused a bit and would like to have some clarification.
Thanks

jvalkeal replied:
Doesn't matter as boot will simply bind to both classes. We don't have shared `MqttProperties` anywhere and didn't want to duplicate same properties.

artembilan replied:
But you did. `Sink` and `Source` variants have many same properties...

jvalkeal replied:
Point was to separate connection properties as what are used in a shared config creating `mqttClientFactory`

artembilan commented on spring-cloud-starter-stream-sink-mqtt/src/main/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkProperties.java:
```diff
@@ -0,0 +1,106 @@
+	/**
+	 * identifies the client
+	 */
+	private String clientId = "scdf.mqtt.client.id.sink";
```
I still don't understand why do we mention SCDF in these apps?
That's good to demonstrate that our products works together and it is really in our interests to promote our products, but these SCSt apps are supposed to be working as a stand-along micorservices or as a part of some other platform.

jvalkeal replied:
When porting from xd, it uses xd in id so followed a pattern. Wonder what it should be?

artembilan commented on spring-cloud-starter-stream-source-mqtt/pom.xml:
```diff
@@ -0,0 +1,61 @@
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.integration</groupId>
+			<artifactId>spring-integration-java-dsl</artifactId>
```
This dependency is out of use.

jvalkeal replied:
Yeah, that was brought over from rabbit app, wonder why it's there?

artembilan replied:
Yep! We don't need it there as well. Good catch!

artembilan reviewed (changes requested):
That's all from my side.
Nothing critical at a glance.
Just some concerns which require a bit more clarification.

jvalkeal committed:
Address review comments#

- Use missing async option
- Added few validations
- Polish pom and @Configuration
- Changed client-id to be max 23

artembilan commented on mqtt-app-starters-common/src/main/java/org/springframework/cloud/stream/app/mqtt/MqttProperties.java:
```diff
@@ -0,0 +1,135 @@
+	 */
+	private String persistenceDirectory = "/tmp/paho";
+
+	@Size(min = 1)
```
Adding validation annotations requires in the latest Boot to add `@Validated` to the class level: https://github.com/spring-cloud-stream-app-starters/app-starters-release/issues/82

jvalkeal replied:
I can add it but weird as I got validation errors without it.

artembilan replied:
@jvalkeal , I pointed you a GH issue originated this request.
Not my call to do that. That is a warn from Boot.

jvalkeal replied:
ah, warning. ok I'll add it.

artembilan commented on spring-cloud-starter-stream-sink-mqtt/src/test/java/org/springframework/cloud/stream/app/mqtt/sink/MqttSinkTests.java:
```diff
@@ -0,0 +1,106 @@
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.cloud.stream.app.test.mqtt.MqttTestSupport;
+//import org.springframework.cloud.stream.binder.test.junit.rabbit.RabbitTestSupport;
```
Redundant code

artembilan reviewed (changes requested):
I don't have anything to nit-pick any more.

jvalkeal committed:
Address review comments

- Add @Validated
- Polish

artembilan commented on mqtt-app-dependencies/pom.xml:
```diff
@@ -0,0 +1,103 @@
+	<modelVersion>4.0.0</modelVersion>
+	<groupId>org.springframework.cloud.stream.app</groupId>
+	<artifactId>mqtt-app-dependencies</artifactId>
+	<version>1.2.1.BUILD-SNAPSHOT</version>
```
There is a discussion on the Tensorflow PR that new app starter should have version as `1.0.0`.

@sobychacko , confirm, please

jvalkeal replied:
I still strongly disagree going with with any other version other than 1.2.x as all others are 1.2.x as well. I would just be confusing.

artembilan reviewed (approved):
I don't have objections on this PR any more.
Can be merged.

Tested via RabbitMQ MQTT plugin. 
I published message to the subscription queue and get answer in the artificially bound `foo` queue like:
```
Exchange 	output
Routing Key 	output
Redelivered 	○
Properties 	
priority:	0
delivery_mode:	2
headers:	
mqtt_retained:	false
mqtt_qos:	0
mqtt_topic:	mqtt-subscription-stream/client/id/sourceqos0
mqtt_duplicate:	false
contentType:	text/plain
content_type:	text/plain
Payload
60 bytes
Encoding: string
	

rty fghjkmhgjkm hgjk hgjkh gjkhgjk ghjk ghjk ghkj ghjk hgj k
```

Tested `sink` there as well via `amqp.topic` and `foo` queue bound via `stream.mqtt` routing key:
```
Exchange 	amq.topic
Routing Key 	stream.mqtt
Redelivered 	○
Properties 	
delivery_mode:	2
headers:	
x-mqtt-publish-qos:	1
x-mqtt-dup:	false
Payload
13 bytes
Encoding: string
	

fghfghfghfgdh
```

garyrussell commented:
Merged as fbc87b2bf542265eb657a59386dcd0738d57e2bf with version 1.0.0, some minor checkstyle changes and switch to use AssertJ.
